### ISSUE TITLE
Crash on Xbox Series X when trying to play any 4k video

### ISSUE THREAD

**Issue by Lake292** [xbmc/xbmc#19044]
### Describe the bug

When trying to play any 4k video on Xbox Series X, it will crash and return to dashboard screen. There are multiple people reporting the same problem on forum here https://forum.kodi.tv/showthread.php?tid=358338.

## Expected Behavior

4k video plays as any other video without crashing the application

## Actual Behavior

When running 4k video, does not have to be HDR it will crash.

## Possible Fix

Not sure, tried to fiddle with settings like adjust refresh rate to always, trying different rendering methods, disabling hardware acceleration without any effect. It always crashed.
Same 4k video works fine in VLC on same system. 

### To Reproduce

1.  Start Kodi on Xbox Series X
2.  Play any 4k video

### Debuglog

The debuglog can be found here:

Here is debug log from the forum https://paste.kodi.tv/awicunilev.kodi, I can provide additional one if needed.

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Windows
 - [X] Windows UWP

 - Operating system version/name: Xbox Series X
 - Kodi version: 18.9

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by lrusak** [comment]
Possible duplicate of #14709

**Comment by Lake292** [comment]
Might be related, though that case describes playback issues. I was able to play 4k videos on Xbox One S, if it was HEVC or too high bitrate it might have issues, but at least it was running. On Xbox Series X it just crashes immediately.
VLC or native Xbox player have no issues with playback of same file.

**Comment by fritsch** [comment]
That issue was Linked cause it containts testbuilds which REVERT the SDK change.

Please don't use issue reports as a "but, but, but I have another personal issue on top".

**Comment by funkmeister7** [comment]
I too encounter this problem.  In the log I have posted [here](https://paste.kodi.tv/musafexeyo.kodi), errors seem to relate to the hardware accelerator.  Disabling the hardware accelerator option crashes Kodi immediately on playing the movie.  With the option enabled, Kodi is playing the file, the audio is coming through but there is no video (the [log I have posted](https://paste.kodi.tv/musafexeyo.kodi) is this second scenario).

I have tried different rippers to create the movie file (mkv) and accessed the files from different paths (i.e. nfs, smb, ftp).  I can play the files using the Xbox player, so the codecs on the box itself are available and working (my assumption Kodi uses these).

**Comment by shAd09** [comment]
I have exactly the same issue on my xbox series s.

**Comment by sbischl** [comment]
tl:dr Kodi is running out of RAM when playing 4k HEVC files.

I did some testing on my Series X using the Dev mode. The issue is still present in Kodi 19 RC1. All 4k HEVC files I tried caused Kodi to crash.  A 4k h264 file played fine using software decoding. I am pretty sure that the HEVC playback issue is caused by Kodi running out of memory. In the Xbox dev mode you can run Kodi as either a normal app or as a game. The latter increases the available RAM from 1 to 4 GB. When declaring Kodi a game, all playback issues are fixed. There have also been log entries which suggest that Kodi is running  out of RAM when constrained to 1GB of RAM. Some research on the forums and on Github suggest that lowering the number of decoder surfaces and swap chain buffers could reduce the ram usage. I set the buffer size to 3 and the number of decoder surfaces to 12. That way I got some 4k HEVC files (seemingly those with lower bitrates) to play. Other files still crash Kodi every time. Further reducing the buffer size or the decoder surfaces introduced new playback issues. I think that to really fix 4k HEVC playback someone has to come up with a way of significantly reducing the RAM usage of 4k HEVC decoding as everything works fine when more memory is available.

**Comment by thexai** [comment]
Try this build:
https://1drv.ms/u/s!Ah8EtIw8AetnhlBKSh3LpUebi2hs?e=pbrduy

I changed some things and is used API `MemoryManager::TrySetAppMemoryUsageLimit` to increase memory available to Kodi but i don't know if is supported on XBOX...

The code is here: 
https://github.com/thexai/xbmc/commit/89340bb98507a16ab5c80a0fcb09a6f9eb4f1133

In any case post full debug log, this will help me to improve this patch.

**Comment by sbischl** [comment]
I tried @thexai 's build. Thanks for trying to fix this! Unfortunatley, the situation hasn't really changed. When running as an App (which gives us 1GB of RAM according to Microsoft's documentation) most HEVC 4k files crash. Logfile of that happening is [here](https://paste.kodi.tv/zanirukila.kodi).  Some low bitrate files work.  When running as a game, we get more RAM and 4k HEVC playback is fixed. [Logfile](https://paste.kodi.tv/uteqihabiz.kodi) of flawless playback when running as game.

Also, I found out why this issue only affects the new Xboxes. There is less RAM available for System + Apps. On Xbox One 5/8GB are reserved for Games. On Series X it's 13.5/16GB. So 500 MB less. We can see the effect of this within Kodi. The System Information screen reports 1536MB of RAM on Series X and 2336MB on Xbox One.

**Comment by thexai** [comment]
Unfortunately Memory Manager API seems not works on Xbox:
```
INFO <general>: DXVA::CContext::CreateContext: [XBOX] App memory usage: current: 375 MB | limit: 1024 | expected limit: 1024 MB
INFO <general>: DXVA::CContext::CreateContext: [XBOX] Try set App memory usage limit to 1960 MB...
ERROR <general>: DXVA::CContext::CreateContext: [XBOX] TrySetAppMemoryUsageLimit failed.

```

Then probably this can only fixed at app / store manifest, declaring Kodi as game for use entire potential.

**Comment by lordofwar81** [comment]
Any updates on this? Brutal not being able to watch any 4k on kodi

**Comment by HarryJaval** [comment]
Another Series X user waiting for Kodi answers..
Maybe when we outnumber on a few months they consider us...nowadays we are a few millions and the Covid is not helping anybody's...not much series X manufactured not much solded vs One,Ps4 users ... patience..

**Comment by Obi-Wan-Aloni** [comment]
With the above comment said, we do appreciate your efforts very much..!
Is there anything that us, the Kodi- Xbox Series X users can do to help expedite a fix to this issue?

**Comment by fuzzard** [comment]
Find a dev interested/with an xbox. No current active team member has an xbox.

**Comment by Obi-Wan-Aloni** [comment]
Bro, I can start fund raising to get an existing developer an Xbox Series X if in return they can prove they're legit and commit to working on a fix quickly once getting it.

Fund-wise, It'll be about 10 USD$ from 50-60 ppl and we're there. Hit the like button if you'll be willing to chip in and post about this suggestion in your local forums.

I know some developers are reading this too so lemme know if someone's interested and let's make that happen!

**Comment by fuzzard** [comment]
Hardware isn't the issue, again, find a DEV interested. There aren't any active team members interested in Xbox (or core windows for that matter) that have either the time or knowledge. 
We are all volunteers who contribute to what interests us, and as has been in blog posts in the past year,  unfortunately we are extremely short on team Devs who are interested in the windows platform.

**Comment by Obi-Wan-Aloni** [comment]
Understood. Will post on some of the communities I'm a part of. Hopefully we can help find someone.

**Comment by tunip3** [comment]
I cannot commit to releasing a fix but I know of a potential solution and will test it to see if it is viable.

**Comment by chivigoku** [comment]
Looking forward to knowing you fix 

here another guy with 4K issues

**Comment by tunip3** [comment]
ok, it appears that the fix worked for my tester, I will open a pr with the fix. The fix is adding the expanded resources rescap to enable more ram to be accessed

**Comment by lordofwar81** [comment]
> ok, it appears that the fix worked for my tester, I will open a pr with the fix. The fix is adding the expanded resources rescap to enable more ram to be accessed

Lemme know when this is live. I'll send u some bitcoin via lightning ✊🏾✊🏾✊🏾

**Comment by tunip3** [comment]
I have pushed my PR

**Comment by tunip3** [comment]
I have pushed my PR, hopefully it gets merged and the app gets updated

**Comment by spydermahn1234** [comment]
Have tested this fix -

used this file on usb2.0 plugged into series x

https://jell.yfish.us/media/jellyfish-200-mbps-4k-uhd-hevc-10bit.mkv

On retail mode with Kodi 18.9 - file makes kodi crash to desktop.

On Dev mode with Kodi 19.0 with expanded resources fix - kodi launched as app (rather than game which allows extra ram by default) and the file plays perfectly.

can provide additional logs if needs be

thanks tunip3 - hope this can get merged asap! great work!

**Comment by spydermahn1234** [comment]
> Hardware isn't the issue, again, find a DEV interested. There aren't any active team members interested in Xbox (or core windows for that matter) that have either the time or knowledge.
> We are all volunteers who contribute to what interests us, and as has been in blog posts in the past year, unfortunately we are extremely short on team Devs who are interested in the windows platform.

Fuzzard - can you look at this commit?? - it appears to have been declined due to what would be a slightly longer wait period for store approval.

It seems a little confusing to me ( I am merely an end user) that when it is stated that Kodi has no active team members interested in Xbox or Core Windows development - that what apparently seems to be a valid fix for an issue that will affect more users in the future is disregarded due to what would be a longer wait for approval by Microsoft.

If you don't have the developers to work on memory reduction - why turn down the efforts of someone who has found a fix/workaround? If at the very least it gets turned down by Microsoft for approval.... Then you could just revert the commit and we are back to where we are now? Doesn't it seem worthwhile to at least try for approval?

**Comment by fuzzard** [comment]
Read the document linked carefully. To use restricted capabilities requires a separate application to Microsoft that they may, or may not allow, at their discretion. 
Therefore merging said "fix" may have the result of Microsoft choosing to reject any further app store submissions of Kodi.
As you say your not a Dev, maybe when the original Dev for Kodi on Xbox says something, you might listen.

An application may be attempted by the team members who do that, but blindly merging something that may result in no app store submission being possible a day out from release must surely be recognised as the wrong time.

The fix is there, there's nothing stopping you from sideloading it yourself. Maybe that's your best option for now.

**Comment by tunip3** [comment]
@fuzzard in that case do you think it would be a good idea to close my pull request?

**Comment by fuzzard** [comment]
No I don't at this point in time.

There is a possibility it can be accepted by Microsoft. If we can get that approval from them, I don't see a reason why we wouldnt merge it at that point. 

Processes need to be done to get to that point though, and it ain't going to happen over night.

**Comment by spydermahn1234** [comment]
> Read the document linked carefully. To use restricted capabilities requires a separate application to Microsoft that they may, or may not allow, at their discretion.
> Therefore merging said "fix" may have the result of Microsoft choosing to reject any further app store submissions of Kodi.
> As you say your not a Dev, maybe when the original Dev for Kodi on Xbox says something, you might listen.
> 
> An application may be attempted by the team members who do that, but blindly merging something that may result in no app store submission being possible a day out from release must surely be recognised as the wrong time.
> 
> The fix is there, there's nothing stopping you from sideloading it yourself. Maybe that's your best option for now.

Good afternoon, please firstly accept my apologies if you thought there was any condescension or bitterness in my original message - it at no point was meant to come across that way.

I have read the document linked carefully - i could not find anywhere in the document that states that Microsoft would penalise Kodi Foundation for attempting an app store submission with restricted capabilities - and as such in the way that i read it, they allow you the option of submitting a new package without the capability request. 

I understand if the timing could be wrong and that is information that i am obviously not privy to.

i also understand that sideloading is an option - however as a father of a young family - getting them to navigate to dev mode really isn't an option for me - First world problems and all that - ("why would your kids want to watch anything in 4k?")

Thank you for your prompt reply - i will listen to you guys - as you know best and have shown that in the over 10 years that i have used xbmc/kodi

**Comment by Chicopower** [comment]
Big thx to everyone working for a better experience with Kodi on Xbox Series.
I enjoy Kodi on my Xbox Series X 👍 

I realy hope Microsoft will accept this kind of "expandedResources" exception or may provide any other kind of exception to go above the 1 GB ram memory limit for Kodi.

Even the default Xbox "Media Player" from Microsoft doesnt support playing my 4k videos.
A so beautifull open source project like Kodi is something that make Xbox more seductive.

Please, Microsoft, if you read this, make the magic happen! :)

**Comment by insalada** [comment]
It's been a while since Microsoft launched Series X. I hope they help out with this as Kodi was my main player for movies on One X and now I cannot longer use it...such a pitty.

**Comment by Obi-Wan-Aloni** [comment]
Hey team. Again, with much appreciation, any update on this issue? Any chance someone can give us a snapshot about where we stand?

**Comment by DaveTBlake** [comment]
Unfortunately the team member that originally got Kodi working on Xbox, and therefore knows most about it, is busy in real life. There is no one else on the active team with a real interest in it and time to spare, in fact we are scrabbling around to try and get v19.0 even tested on Xbox. Hence release on Windows Store has been delayed until that has happened, as we don't want to risk having users auto update to untested software.

We are aware that there is an issue with 4k, and that someone has even raised a PR to fix https://github.com/xbmc/xbmc/pull/19221. However there are problems with that fix and Windows Store submissions, so a better solution needs to be found.

As has been said before, it isn't that no one cares just that we do not have the people with the knowledge and spare time to volunteer to deal with this.

**Comment by Obi-Wan-Aloni** [comment]
Are there any updates?
I'm asking, being aware of the reply above me, as I've noticed Kodi 19 is now available on the Xbox Store.

**Comment by DaveTBlake** [comment]
Yes v19.0 is on Windows Store but for Windows 10 only, **not for Xbox**.

**Comment by Obi-Wan-Aloni** [comment]
You're right. I'm Wrong. I apologize :)

I was under an assumption I guess.. Since I recently got a notification when launching Kodi on my Xbox that there's a new version available. However, entering the Xbox store now I couldn't see an option to update the app.

**Comment by HitcherUK** [comment]
The Kodi update available app isn't platform specific unfortunately hence why you're seeing the notification.

**Comment by dudzio12** [comment]
> Yes v19.0 is on Windows Store but for Windows 10 only, **not for Xbox**.

You said earlier that you did not want to release untested software, you've managed to test it for Windows platform. So what is the issue with Xbox when you keep in mind that it is not currently working as expected? 🤔

**Comment by thexai** [comment]
In the next few days I will try to solve some problems that has UWP (Windows) version currently and are also common to XBOX. There are other issues like this about 4K that can affect XBOX only.

Right now the UWP versions (Windows and XBOX) are broken because they don't have Internet connectivity. It seems to be related to "recent" changes in cURL or python or SSL certificates. But I haven't really been able to spend time on it other than looking at recent git history, kodi logs posted etc. This problem a few months ago did not exist.

The thing about UWP is that since there is no developer who uses it, the problems that arise (runtime), nobody notices until after a long time.
